### PR TITLE
LoggingTransport

### LINKED ISSUE #401: Log RPC calls to Node
In GPv1 it helped a lot when debugging issues with the nodes to have debug logs of all RPC requests and responses.

Today, I was debugging a "lost transaction" issue and wasn't able to see the response we received from e.g. the `eth_sendRawTransaction`.

I think we should somehow reuse our custom HTTP implementation from v1 that logs requests.

### PR DESCRIPTION
**Pull request by fleupold**
Fixes #401

This PR implements a custom (Batch)Transport implementation that logs all requests and responses for debugging purposes.

I tried to use the DynTransport trait inside the call sites instead of comitting to a concrete implementation. One issue is that DynTransport is not implemented for BatchTransport. I have a ethcontracts PR ready that adds `DynBatchTransport`, however after discussing with @nlordell we decided that it might be simpler to just commit to a concrete type in one place (shared/lib.rs) and then use this type on all callsites. We can make it more generic if different callsites have different transport requirements (e.g. in unit tests we might not want to build all our Transport layers)

### Test Plan
Run e.g. the solver binary.

See RPC requests being logged like this:

```
2021-03-31T15:54:46.971Z DEBUG shared::transport: [id:6] sending request: '{"jsonrpc":"2.0","method":"net_version","params":[],"id":6}'
2021-03-31T15:54:47.072Z DEBUG shared::transport: [id:6] received response: '"4"'

[...]

2021-03-31T15:54:47.775Z DEBUG shared::transport: [batch_id:19] sending Batch:
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x0902f1ac","to":"0xa913d1f89b96215ec62e039fb6bf7b30846fb773"},"latest"],"id":19}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a08231000000000000000000000000a913d1f89b96215ec62e039fb6bf7b30846fb773","to":"0x6e894660985207feb7cf89faf048998c71e8ee89"},"latest"],"id":20}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a08231000000000000000000000000a913d1f89b96215ec62e039fb6bf7b30846fb773","to":"0xc778417e063141139fce010982780140aa0cd5ab"},"latest"],"id":21}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x0902f1ac","to":"0x57a0611e5ba78564d4cc37ce8424f37ab4cec771"},"latest"],"id":22}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a0823100000000000000000000000057a0611e5ba78564d4cc37ce8424f37ab4cec771","to":"0x577d296678535e4903d59a4c929b718e1d575e0a"},"latest"],"id":23}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a0823100000000000000000000000057a0611e5ba78564d4cc37ce8424f37ab4cec771","to":"0x6e894660985207feb7cf89faf048998c71e8ee89"},"latest"],"id":24}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x0902f1ac","to":"0x937959382d811bd84463b3d1d221fa093930dd05"},"latest"],"id":25}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a08231000000000000000000000000937959382d811bd84463b3d1d221fa093930dd05","to":"0x577d296678535e4903d59a4c929b718e1d575e0a"},"latest"],"id":26}
  {"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x70a08231000000000000000000000000937959382d811bd84463b3d1d221fa093930dd05","to":"0xc778417e063141139fce010982780140aa0cd5ab"},"latest"],"id":27}
2021-03-31T15:54:47.886Z DEBUG shared::transport: [batch_id:19] received response:
  [id:19]: "0x00000000000000000000000000000000000000000000000007ca4c1c009ff5c800000000000000000000000000000000000000000000000007ce4a7b81f16f510000000000000000000000000000000000000000000000000000000060647ed7"
  [id:20]: "0x00000000000000000000000000000000000000000000000007ca4c1c009ff5c8"
  [id:21]: "0x00000000000000000000000000000000000000000000000007ce4a7b81f16f51"
  [id:22]: "0x"
  [id:23]: "0x0000000000000000000000000000000000000000000000000000000000000000"
  [id:24]: "0x0000000000000000000000000000000000000000000000000000000000000000"
  [id:25]: "0x000000000000000000000000000000000000000000000000000004fdec89a9b400000000000000000000000000000000000000000000000008b5b5db5c19fb1f0000000000000000000000000000000000000000000000000000000060643dc3"
  [id:26]: "0x000000000000000000000000000000000000000000000000000004fdec89a9b4"
  [id:27]: "0x00000000000000000000000000000000000000000000000008b5b5db5c19fb1f"
```

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]

File: shared/src/transport.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        self.inner.prepare(method, params)
+    }
+
+    fn send(&self, id: RequestId, request: Call) -> Self::Out {
+        if let Ok(serialized) = serde_json::to_string(&request) {
+            tracing::debug!("[id:{}] sending request: '{}'", id, &serialized);
+        }
+        self.inner
+            .send(id, request)
+            .map(move |response| {
```
Comment (nlordell) [inline shared/src/transport.rs:35]: ubernit: I think [`inspect`](https://docs.rs/futures/0.3.13/futures/future/trait.FutureExt.html#method.inspect) is slightly more appropriate here as it is for:

> Do something with the output of a future before passing it on.

File: shared/src/transport.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            batch_id,
+            requests
+                .iter()
+                .filter_map(|(_, call)| Some(format!("  {}", serde_json::to_string(call).ok()?)))
+                .collect::<Vec<_>>()
+                .join("\n")
+        );
+        self.inner
+            .send_batch(requests.clone())
+            .map(move |response| {
```
Comment (nlordell) [inline shared/src/transport.rs:72]: same as :point_up:

**Review by nlordell — APPROVED** [review]

File: shared/src/transport.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        let requests: Vec<_> = requests.into_iter().collect();
+        // Empty batches are pointless and can therefore have a 0 id, otherwise we use the ID of the first request.
+        let batch_id = requests.first().map(|(id, _)| *id).unwrap_or_default();
+        tracing::debug!(
+            "[batch_id:{}] sending Batch:\n{}",
+            batch_id,
+            requests
+                .iter()
+                .filter_map(|(_, call)| Some(format!("  {}", serde_json::to_string(call).ok()?)))
```
Comment (nlordell) [inline shared/src/transport.rs:66]: maybe print the id here as well to make it easier to line up with the responses?

Comment (fleupold) [reply shared/src/transport.rs:66]: The ID is already part of the request (cf. output in test plan) so I thought it would be redundant.

Comment (nlordell) [reply shared/src/transport.rs:65]: Oh yeah!, its at the very end of the JSON. Sorry didn't see. YOu're right, its completely redundant then.

**Pull request merged**